Omnifoldwbkg

wecassidy opened the pull request:
I want to refactor how `OmniFoldwBkg` handles background events, particularly the no-background case. Basically there's a lot of code that looks like this that I'd like to clean up ([omnifoldwbkg.py:216-226](https://github.com/zctao/omnifoldTop/blob/d2b77d6e57e370fd3f2a44be35e14fb6ff877139/python/omnifoldwbkg.py#L216-L226)):
```python
        X_obs = self.datahandle_obs[self.vars_reco]
        if self.datahandle_obsbkg is not None:
            X_obsbkg = self.datahandle_obsbkg[self.vars_reco]
            X_obs = np.concatenate([X_obs, X_obsbkg])

        X_sig = self.datahandle_sig[self.vars_reco]

        if self.datahandle_bkg is not None:
            X_bkg = self.datahandle_bkg[self.vars_reco]
        else:
            X_bkg = None
```

Two ways to go about it I've been thinking about:

# Separate the background and no-background cases
Maybe at the OmniFold level, so have `OmniFold` (without background) and `OmniFoldwBkg` classes which expose a common interface. Lots of code doesn't care about the distinction between with and without background, though. Lots of code duplication that way without some sort of code-sharing between the two cases. There doesn't seem to be a natural inheritance relation between the two cases, and I think it would be a pretty major effort to get the code to a place where that makes sense.

Could alternatively think about splitting the "handling groups of `DataHandler`s" parts out of `OmniFoldwBkg`. Instead of holding 2-4 datahandlers (obs, sig, bkg, obsbkg) instead `OmniFoldwBkg` has a "`DataHandlerGroup`" (or some other better name). Where `OmniFoldwBkg` currently calls datahandlers, it could instead call the group handler which would somehow transparently handle the presence or absence of background.

# No background is an empty array
In this approach, instead of setting a `self.datahandle_bkg` to `None` if there is no background data, instead I set it to a subclass of `DataHandler` that represents an empty dataset, analogous to an empty data array. Then the code above would become:

```python
        X_obs = self.datahandle_obs[self.vars_reco]
        X_sig = self.datahandle_sig[self.vars_reco]
        X_bkg = self.datahandle_bkg[self.vars_reco]
        X_obsbkg = self.datahandle_obsbkg[self.vars_reco]

        X_obs = np.concatenate([X_obs, X_obsbkg])
```

In the no-background case, `X_obsbkg` is an array of shape (0, n_vars), so concatenating it to `X_obs` has no effect. In the background case, the result is the same as before. I think in most cases in OmniFoldwBkg no special case check would be required to handle the presence or absence of background events.

This approach seems more elegant to me to me than the one above, so I've started exploring it by building the `Empty` class in f856b50.

wecassidy committed:
Make code to construct unfolder more concise

wecassidy committed:
Pull _reweight out of OmniFoldwBkg since it doesn't use `self`

wecassidy committed:
Move model setup and training to model.py

Neither method used OmniFoldwBkg state, so they shouldn't be
OmniFoldwBkg members. These actions conceptually group better with
model code than with OmniFoldwBkg code.

wecassidy committed:
_read_weights_from_file doesn't use self:  move out of OmniFoldwBkg

wecassidy committed:
Empty dataset class. Should make handling an no background cleaner.

zctao commented:
Does `X_obs = np.concatenate([X_obs, X_obsbkg])` make an unnecessary copy of `X_obs` and double the memory usage in the no-background case? 
Many of the if statements as the one you mentioned above were me being paranoid about memory usage when we need to process and train a much larger data sample for the actual analysis.

zctao commented:
Moving the methods for setting up and training models out of `OmniFoldwBkg` looks good. 
I have a pull request #18 that probably has some conflicts with your changes. Do you mind if I merge #18 in first?

wecassidy commented:
Yeah, go ahead with #18. This is very much exploratory, I don't expect to be done with it for at least a week.

zctao commented:
cool thanks!

wecassidy commented:
> Does `X_obs = np.concatenate([X_obs, X_obsbkg])` make an unnecessary copy of `X_obs` and double the memory usage in the no-background case?
> Many of the if statements as the one you mentioned above were me being paranoid about memory usage when we need to process and train a much larger data sample for the actual analysis.

Based on the [`np.concatenate` source](https://github.com/numpy/numpy/blob/3544fae2859d8ecac378b9ddc8f909c69455a831/numpy/core/src/multiarray/multiarraymodule.c#L492-L498), I think it will always copy. The old `X_obs` might get garbage collected quickly, but I'm not sure what guarantees if any Python makes about that.

wecassidy commented:
OK, so most GC in Python is handled by reference counting [1]. So if I'm understanding correctly the old `X_obs` should get deallocated after it's overwritten, but for the duration of `np.concatenate` there will be two copies. That interpretation is supported by this test script:

```python
import tracemalloc
import sys

import numpy as np

tracemalloc.start()

a = np.random.random_sample((1000, 1000, 1000))

first_size = sys.getsizeof(a)
first_total, first_peak = tracemalloc.get_traced_memory()

a = np.concatenate([a])

second_size = sys.getsizeof(a)
second_total, second_peak = tracemalloc.get_traced_memory()

GiB = 2 ** 30

print(f"Size of `a` before concatenate: {first_size / GiB:.2f} GiB")
print(f"Total memory allocated before concatenate: {first_total / GiB:.2f} GiB")
print(f"Peak memory use before concatenate: {first_peak / GiB:.2f} GiB")
print()
print(f"Size of `a` after concatenate: {second_size / GiB:.2f} GiB")
print(f"Total memory allocated after concatenate: {second_total / GiB:.2f} GiB")
print(f"Peak memory use after concatenate: {second_peak / GiB:.2f} GiB")
```
Which has the following output:
```
Size of `a` before concatenate: 7.45 GiB
Total memory allocated before concatenate: 7.45 GiB
Peak memory use before concatenate: 7.45 GiB

Size of `a` after concatenate: 7.45 GiB
Total memory allocated after concatenate: 7.45 GiB
Peak memory use after concatenate: 14.90 GiB
```

Copying large datasets will always be slower than not copying, but is it enough of a hit to matter compared to training time? And are you concerned about running out of memory during the `np.concatenate`, or having the old `X_obs` stick around and leak memory?


[1] https://docs.python.org/3/c-api/intro.html#objects-types-and-reference-counts

zctao commented:
Indeed timing is not really a concern, but out of memory might be. 
I haven't checked exactly how large the data array is for the LHC run 2 data, maybe it's ok.

zctao commented:
Btw, `X_obsbkg` is never needed if we are unfolding the actual data. It is only used when we mix MC samples to create a pseudo dataset for testing the performance on data with non-negligible background events.

wecassidy committed:
Remove background special-casing where it can't affect memory use

This does change the appearance of reco plots slightly: they will
always have a "Bkg." entry in the legend, though the actual background
histogram isn't visible because its elements are all zeros.

wecassidy commented:
5bd6d1f removes the checks that the background datahandler is not `None` where it won't change performance. (TODO: the current check for an absent `DataHandler`, doesn't work since no background value is now `Empty` instead of `None`. I need to remember to fix this.)

There's a change to the appearance of the reco plots in this commit: there will always be a "Bkg." entry in the legend, even if background events aren't provided. You can't actually see the background histogram because all the elements are zero. I could reintroduce the check there if you think it matters?

zctao commented:
> There's a change to the appearance of the reco plots in this commit: there will always be a "Bkg." entry in the legend, even if background events aren't provided. You can't actually see the background histogram because all the elements are zero. I could reintroduce the check there if you think it matters?

It's ok. We can worry about formatting plots later.

wecassidy committed:
Remove special case for background mixed with data

wecassidy committed:
truth_known is only used in plot_distributions_unfold, pass it there

wecassidy committed:
Factor model setup code out of OmniFoldwBkg into model.py

wecassidy committed:
Fix error introduced in c84af91

Calculating metrics no longer checks for truth known

wecassidy committed:
Rearrange functions in model.py into a more logical order

wecassidy committed:
Refactor code related to loading model weights from a file

wecassidy committed:
Fix bug introduced by the fix in 1ee48e4 (sigh)

wecassidy committed:
Refactor label generation in step 1

wecassidy committed:
Factor preprocessing out of _get_feature_arrays_stepN

wecassidy committed:
Factor the step 1 and 2 likelihood ratio calculation into a method

The code is almost identical between the two

wecassidy committed:
Merge branch 'master' into omnifoldwbkg

wecassidy committed:
Merge branch 'master' into omnifoldwbkg

wecassidy commented:
@zctao I don't remember what more I planned to do on this branch. I think it can be merged as-is.

zctao commented:
Thanks! There might be some conflicts with the stuff I have on another branch. I'll take a closer took.
